### PR TITLE
GitHub Actions

### PR DESCRIPTION
**Pull request by nafg**

### CODE REVIEW

**Comment by nafg** [comment]
Fixes #2212

**Comment by nafg** [comment]
Fixes #2211

**Comment by nafg** [comment]
Hmm, it seems like everything is passing now except for https://github.com/slick/slick/runs/2434592197?check_suite_focus=true#step:7:1794:

```
[error] Test com.typesafe.slick.testkit.tests.SequenceTest.test2[mysql] failed: java.sql.SQLException: This function has none of DETERMINISTIC, NO SQL, or READS SQL DATA in its declaration and binary logging is enabled (you *might* want to use the less safe log_bin_trust_function_creators variable), took 0.045 sec
1795
[error]     at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
1796
[error]     at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
1797
[error]     at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:953)
1798
[error]     at com.mysql.cj.jdbc.ClientPreparedStatement.execute(ClientPreparedStatement.java:370)
1799
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.$anonfun$run$5(JdbcActionComponent.scala:287)
1800
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.$anonfun$run$5$adapted(JdbcActionComponent.scala:287)
1801
[error]     at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)
1802
[error]     at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)
1803
[error]     at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)
1804
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.$anonfun$run$4(JdbcActionComponent.scala:287)
1805
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.$anonfun$run$4$adapted(JdbcActionComponent.scala:287)
1806
[error]     at scala.collection.immutable.Vector.foreach(Vector.scala:1856)
1807
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.run(JdbcActionComponent.scala:287)
1808
[error]     at slick.jdbc.JdbcActionComponent$SchemaActionExtensionMethodsImpl$$anon$5.run(JdbcActionComponent.scala:285)
1809
[error]     at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
1810
[error]     at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
1811
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
1812
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
1813
[error]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
1814
[error]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
1815
[error]     at java.lang.Thread.run(Thread.java:831)
```

**Comment by nafg** [comment]
Does not yet support releasing or publishing docs

**Comment by nafg** [comment]
I adapted admin/build.sh so github actions can use it too. Hopefully this doesn't break travis and means that releases can be done via github actions.

What is the main reason for Appveyor? To test more versions of SQL Server? To test SQL Server on Windows (even though it's already tested on Linux)? Something else?

**Comment by SethTisue** [comment]
> What is the main reason for Appveyor?

my dig turned up #1443 which is titled "add Appveyor build for SQLServer". but perhaps also a secondary motivation to have Windows testing?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    branches: [ main, nafg/build-fixes ]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    
+    strategy:
+      matrix:
+        java: [1.8.0, 1.11.0, 1.16.0]
+        scala: [2.12.3, 2.13.5]
```
Comment (SethTisue) [inline .github/workflows/ci.yml:16]: why 2.12.3 rather than 2.12.13?

**Review by SethTisue — COMMENTED** [review]
LGTM generally — I assume the idea is that we'd eventually disable Travis-CI, after we've tried this for a while?

**Comment by nafg** [comment]
> > What is the main reason for Appveyor?
> 
> my dig turned up #1443 which is titled "add Appveyor build for SQLServer". but perhaps also a secondary motivation to have Windows testing?

I mean if it just predates the existence of SQL Server for linux then maybe now it's obsolete. I guess we could search for issues that seem Windows-only.

I mean GitHub Actions supports Windows (and MacOS) but the question is how far to go. Right now appveyor is testing a lot of SQL Server versions, but I don't know what's needed.

I don't know how many jobs GitHub will run in parallel for OSS. I don't mind going all out but I don't want PRs to take longer than necessary to check, even when there are multiple PRs to build.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (nafg) [reply .github/workflows/ci.yml:16]: Typo, good catch

File: build.sbt [hunk 2]
```diff
@@ -147,7 +147,10 @@ lazy val sampleSlickTestkitExampleProject =
 
 /* A command that runs all tests sequentially */
 def testAll = Command.command("testAll") { state =>
+  println("Starting testAll")
   val extracted = Project.extract(state)
+  
+  println("testAll: got state")
```
[the remark this answers is not in the document]
Comment (nafg) [reply build.sbt:153]: No, good catch. Also I would squash the commits. (At least most of them.)

**Comment by nafg** [comment]
> LGTM generally — I assume the idea is that we'd eventually disable Travis-CI, after we've tried this for a while?

Yeah, especially because it's so slow now. It seems to only run one job at a time.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (nafg) [reply .github/workflows/ci.yml:16]:
```suggestion
        scala: [2.12.13, 2.13.5]
```

File: build.sbt [hunk 3]
```diff
@@ -147,7 +147,10 @@ lazy val sampleSlickTestkitExampleProject =
 
 /* A command that runs all tests sequentially */
 def testAll = Command.command("testAll") { state =>
+  println("Starting testAll")
```
Comment (nafg) [inline build.sbt:150]:
```suggestion
```

File: build.sbt (hunk 2 above)
Comment (nafg) [inline build.sbt:153]:
```suggestion
```

**Comment by nafg** [comment]
I'm thinking to DRY the travis and testkit config files using [`include`](https://github.com/lightbend/config#:~:text=include%20feature%20merges%20root%20object%20in%20another%20file%20into%20current%20object%2C%20so%20foo%20%7B%20include%20%22bar.json%22%20%7D%20merges%20keys%20in%20bar.json%20into%20the%20object%20foo)

**Comment by nafg** [comment]
@SethTisue @octonato do people use Slick with DB2? I think it's a commercial database that was probably originally in Slick Extensions.

For some reason Travis seems to succeed even though db2 is enabled and I don't think the jar is copied into the right place, since travis logs often have a db2 ClassNotFoundException. I changed the location the jar is copied to to slick-testkit/lib which means it's picked up, but now there's some kind of failure:

```
com.ibm.db2.jcc.am.DisconnectNonTransientConnectionException: [jcc][t4][2030][11211][4.28.11] A communication error occurred during operations on the connection's underlying socket, socket input stream, 
or socket output stream.  Error location: Reply.fill() - insufficient data (-1).  Message: Insufficient data. ERRORCODE=-4499, SQLSTATE=08001
```

It's possible it's something trivial, or not...

**Comment by nafg** [comment]
Most issues are solved:

- [x] Releasing and docs (reuses admin/build.sh)
- [x] DRYed the HOCON wrt. Travis
- [x] DB2 tests are working (the container needed --privileged)

What's unresolved are the following questions:
1. What tests should run on Windows?
2. What SQL Server version(s) should be tested?

Note that there are tests defined for certain versions, although other than the name I'm not sure how different the tests are.

See https://github.com/slick/slick/blob/main/slick-testkit/src/test/scala/slick/test/profile/ProfileTest.scala#L49-L70 

and https://github.com/slick/slick/blob/main/slick-testkit/src/main/scala/com/typesafe/slick/testkit/util/StandardTestDBs.scala#L189-L235

**Comment by nafg** [comment]
I squashed it all into two commits, one that changes tabs to spaces in runcontainer.sh, and one that is the actual changes. I'm ready to merge it, it just needs a review for that to be possible.

**Comment by SethTisue** [comment]
I have no clue how widely used DB2 is. If you have reason to believe it isn't used much, then I think it's fine to either disable the testing or even remove the support entirely, as long as you open an issue saying that we're doing so and that if somebody wants it back, they need to volunteer to do the work of bringing it back. Then when we do a release candidate, the release notes would have a link to that issue, and either a volunteer emerges or one doesn't.

**Review by SethTisue — APPROVED** [review]

**Comment by nafg** [comment]
> I have no clue how widely used DB2 is. If you have reason to believe it isn't used much, then I think it's fine to either disable the testing or even remove the support entirely, as long as you open an issue saying that we're doing so and that if somebody wants it back, they need to volunteer to do the work of bringing it back. Then when we do a release candidate, the release notes would have a link to that issue, and either a volunteer emerges or one doesn't.

@SethTisue I got db2 working. Thanks for reviewing

**Pull request merged**